compiler/rustc_lint_defs/src/builtin.rs

```diff
@@ -4380,3 +4381,53 @@ declare_lint! {
     "effective visibility of a type is larger than the area in which it can be named",
     @feature_gate = sym::type_privacy_lints;
 }
+
+declare_lint! {
+    /// The `coinductive_overlap_in_coherence` lint detects impls which are currently
+    /// considered not overlapping, but may be considered to overlap if support for
+    /// coinduction is added to the trait solver.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,compile_fail
+    /// use std::borrow::Borrow;
+    /// use std::cmp::Ordering;
+    /// use std::marker::PhantomData;
+    ///
+    /// #[derive(PartialEq, Default)]
+    /// pub(crate) struct Interval<T>(T);
+
+    /// impl<T, Q> PartialEq<Q> for Interval<T>
+    /// where
+    ///     Q: PartialOrd,
+    /// {
+    ///     fn eq(&self, other: &Q) -> bool {
+    ///         todo!()
+    ///     }
+    /// }
+    ///
+    /// impl<T, Q> PartialOrd<Q> for Interval<T>
+    /// where
+    ///     Q: PartialOrd,
+    /// {
+    ///     fn partial_cmp(&self, other: &Q) -> Option<Ordering> {
+    ///         todo!()
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// The manual impl of `PartialEq` impl overlaps with the `derive`, since
+    /// if we replace `Q = Interval<T>`, then the second impl leads to a cycle:
+    /// `PartialOrd for Interval<T> where Interval<T>: Partial`.
```

Comment:
I would repeat the above here again:
```suggestion
    /// `PartialOrd for Interval<T> where Interval<T>: PartialOrd`. This cycle
    /// currently causes the compiler to consider `Interval<T>: PartialOrd` to not
    /// hold, causing the two implementations to be disjoint. This will change in
    /// a future release.
```